How to construct a font from a memory address?

To avoid downloading a large number of fonts every time I update the firmware, I first upload the fonts to a fixed FLASH location. This way, I can share the fonts across multiple programs

```
error[E0080]: memory access failed: attempting to access 1 byte, but got 0xce000[noalloc] which is a dangling pointer (it has no provenance)
  --> src/globals.rs:55:35
   |
55 | pub static WQY16: FontRenderer  = FontRenderer::new_from_slice(core::slice::from_raw_parts(0xce000 as *const u8, 24576));
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `globals::WQY16` failed inside this call
   |
note: inside `FontRenderer::new_from_slice`
  --> /home/lei/code/github/u8g2-fonts/src/renderer.rs:46:19
   |
46 |             font: FontReader::new_from_slice(font_data),
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `u8g2_fonts::font_reader::FontReader::new_from_slice`
  --> /home/lei/code/github/u8g2-fonts/src/font_reader/mod.rs:41:26
   |
41 |             glyph_count: data[0],
   |                          ^^^^^^^ the failure occurred here


```
This cannot be compiled

Note that this fails (that's exactly your error): 
```rust
const fn bar() -> u8 {
    unsafe {
        let ptr = 0x1234 as *const u8;
        let slice = core::slice::from_raw_parts(ptr, 100);

        slice[0]
    }
}

static foo: u8 = bar();
```
```
error[E0080]: memory access failed: attempting to access 1 byte, but got 0x1234[noalloc] which is a dangling pointer (it has no provenance)
  --> src/lib.rs:11:18
   |
11 | static foo: u8 = bar();
   |                  ^^^^^ evaluation of `foo` failed inside this call
   |
note: inside `bar`
  --> src/lib.rs:7:9
   |
 7 |         slice[0]
   |         ^^^^^^^^ the failure occurred here

For more information about this error, try `rustc --explain E0080`.
```

While this works:
```rust
use lazy_static::lazy_static;

const fn bar() -> u8 {
    unsafe {
        let ptr = 0x1234 as *const u8;
        let slice = core::slice::from_raw_parts(ptr, 100);

        slice[0]
    }
}

lazy_static! {
    static ref foo: u8 = bar();
}

fn main() {
    println!("{}", *foo);
}
```

Although it segfaults, of course.

AI informs me that `const` does not support `lazy_static`. However,
everything in this library is `const`. The `const` keyword requires that
content be determined at compile time, whereas loading data from a Flash
address necessitates waiting until runtime.

Finomnis ***@***.***> 于 2026年5月24日周日 23:57写道：

> *Finomnis* left a comment (Finomnis/u8g2-fonts#41)
> <https://github.com/Finomnis/u8g2-fonts/issues/41#issuecomment-4529245935>
>
> Note that this fails (that's exactly your error):
>
> const fn bar() -> u8 {
>     unsafe {
>         let ptr = 0x1234 as *const u8;
>         let slice = core::slice::from_raw_parts(ptr, 100);
>
>         slice[0]
>     }}
> static foo: u8 = bar();
>
> error[E0080]: memory access failed: attempting to access 1 byte, but got 0x1234[noalloc] which is a dangling pointer (it has no provenance)
>   --> src/lib.rs:11:18
>    |
> 11 | static foo: u8 = bar();
>    |                  ^^^^^ evaluation of `foo` failed inside this call
>    |
> note: inside `bar`
>   --> src/lib.rs:7:9
>    |
>  7 |         slice[0]
>    |         ^^^^^^^^ the failure occurred here
>
> For more information about this error, try `rustc --explain E0080`.
>
> While this works:
>
> use lazy_static::lazy_static;
> const fn bar() -> u8 {
>     unsafe {
>         let ptr = 0x1234 as *const u8;
>         let slice = core::slice::from_raw_parts(ptr, 100);
>
>         slice[0]
>     }}
> lazy_static! {
>     static ref foo: u8 = bar();}
> fn main() {
>     println!("{}", *foo);}
>
> Although it segfaults, of course.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Finomnis/u8g2-fonts/issues/41#issuecomment-4529245935>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALBRX5QNZMWLHCLG5ODL4344MLYJAVCNFSM6AAAAACZK5MF46VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DKMRZGI2DKOJTGU>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

--- a/src/font_reader/mod.rs
+++ b/src/font_reader/mod.rs
@@ -7,7 +7,7 @@ mod glyph_renderer;
 mod glyph_searcher;
 mod unicode_jumptable_reader;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Eq, PartialEq)]
 pub struct FontReader {
     pub data: DebugIgnore<&'static [u8]>,
     pub supports_background_color: bool,

--- a/src/font_reader/mod.rs
+++ b/src/font_reader/mod.rs
@@ -35,7 +35,7 @@ pub struct FontReader {
 }
 
 impl FontReader {
-    pub const fn new_from_slice(data: &'static [u8]) -> Self {
+    pub const fn new_from_raw_data(data: &'static [u8]) -> Self {
         let mut this = Self {
             data: DebugIgnore(data),
             glyph_count: data[0],
@@ -67,7 +67,7 @@ impl FontReader {
 
     pub const fn new<F: Font>() -> Self {
         let data = F::DATA;
-        Self::new_from_slice(data)
+        Self::new_from_raw_data(data)
     }
 
     pub const fn with_ignore_unknown_glyphs(mut self, ignore: bool) -> Self {

--- a/src/font_reader/mod.rs
+++ b/src/font_reader/mod.rs
@@ -35,9 +35,7 @@ pub struct FontReader {
 }
 
 impl FontReader {
-    pub const fn new<F: Font>() -> Self {
-        let data = F::DATA;
-
+    pub const fn new_from_slice(data: &'static [u8]) -> Self {
         let mut this = Self {
             data: DebugIgnore(data),
             glyph_count: data[0],
@@ -65,6 +63,11 @@ impl FontReader {
         };
         this.line_height = this.get_default_line_height() as u32;
         this
+    }
+
+    pub const fn new<F: Font>() -> Self {
+        let data = F::DATA;
+        Self::new_from_slice(data)
     }
 
     pub const fn with_ignore_unknown_glyphs(mut self, ignore: bool) -> Self {

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -397,4 +397,12 @@ mod tests {
             FontRenderer::new::<crate::fonts::u8g2_font_u8glib_4_tf>()
         );
     }
+
+    #[test]
+    fn new_from_slice() {
+        let renderer_from_font = FontRenderer::new::<crate::fonts::u8g2_font_u8glib_4_tf>();
+        let renderer_from_slice =
+            FontRenderer::new_from_slice(crate::fonts::u8g2_font_u8glib_4_tf::DATA);
+        assert_eq!(renderer_from_font.font, renderer_from_slice.font);
+    }
 }

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -408,6 +408,6 @@ mod tests {
         let renderer_from_font = FontRenderer::new::<crate::fonts::u8g2_font_u8glib_4_tf>();
         let renderer_from_raw_data =
             FontRenderer::new_from_raw_data(crate::fonts::u8g2_font_u8glib_4_tf::DATA);
-        assert_eq!(renderer_from_font.font, renderer_from_slice.font);
+        assert_eq!(renderer_from_font.font, renderer_from_raw_data.font);
     }
 }

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -36,6 +36,17 @@ impl FontRenderer {
         }
     }
 
+    /// Creates a new instance of a font renderer from custom font data.
+    ///
+    /// # Arguments
+    ///
+    /// * `font_data` - the font to render. Must adhere to the [U8g2 font format](https://github.com/olikraus/u8g2/wiki/u8g2fontformat).
+    pub const fn new_from_slice(font_data: &'static [u8]) -> Self {
+        Self {
+            font: FontReader::new_from_slice(font_data),
+        }
+    }
+
     /// Switches the font rendering mode to ignore all unrenderable characters
     /// instead of raising an error.
     ///

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -41,6 +41,11 @@ impl FontRenderer {
     /// # Arguments
     ///
     /// * `font_data` - the font to render. Must adhere to the [U8g2 font format](https://github.com/olikraus/u8g2/wiki/u8g2fontformat).
+    ///
+    /// # Panics
+    ///
+    /// Panics if `font_data` is not valid U8g2 font data, including if it is
+    /// shorter than the required U8g2 font header.
     pub const fn new_from_slice(font_data: &'static [u8]) -> Self {
         Self {
             font: FontReader::new_from_slice(font_data),

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -46,9 +46,9 @@ impl FontRenderer {
     ///
     /// Panics if `font_data` is not valid U8g2 font data, including if it is
     /// shorter than the required U8g2 font header.
-    pub const fn new_from_slice(font_data: &'static [u8]) -> Self {
+    pub const fn new_from_raw_data(font_data: &'static [u8]) -> Self {
         Self {
-            font: FontReader::new_from_slice(font_data),
+            font: FontReader::new_from_raw_data(font_data),
         }
     }
 
@@ -404,10 +404,10 @@ mod tests {
     }
 
     #[test]
-    fn new_from_slice() {
+    fn new_from_raw_data() {
         let renderer_from_font = FontRenderer::new::<crate::fonts::u8g2_font_u8glib_4_tf>();
         let renderer_from_slice =
-            FontRenderer::new_from_slice(crate::fonts::u8g2_font_u8glib_4_tf::DATA);
+            FontRenderer::new_from_raw_data(crate::fonts::u8g2_font_u8glib_4_tf::DATA);
         assert_eq!(renderer_from_font.font, renderer_from_slice.font);
     }
 }

--- a/src/renderer.rs
+++ b/src/renderer.rs
@@ -406,7 +406,7 @@ mod tests {
     #[test]
     fn new_from_raw_data() {
         let renderer_from_font = FontRenderer::new::<crate::fonts::u8g2_font_u8glib_4_tf>();
-        let renderer_from_slice =
+        let renderer_from_raw_data =
             FontRenderer::new_from_raw_data(crate::fonts::u8g2_font_u8glib_4_tf::DATA);
         assert_eq!(renderer_from_font.font, renderer_from_slice.font);
     }

--- a/src/utils/debug_ignore.rs
+++ b/src/utils/debug_ignore.rs
@@ -1,6 +1,6 @@
 use core::ops::{Deref, DerefMut};
 
-#[derive(Clone)]
+#[derive(Clone, Eq, PartialEq)]
 pub struct DebugIgnore<T>(pub T);
 
 impl<T> core::fmt::Debug for DebugIgnore<T> {

